lib/spack/spack/package.py

```diff
@@ -1587,6 +1593,15 @@ def do_install(self, **kwargs):
         their build process.
 
         Args:"""
+        # Non-transitive dev specs need to keep the dev stage and be built from
+        # source every time. Transitive ones just need to be built from source.
+        dev_path_var = self.spec.variants.get('dev_path', None)
+        if dev_path_var:
+            kwargs['keep_stage'] = True
+        dev_build_var = self.spec.variants.get('dev_build', None)
+        if dev_build_var:
+            kwargs['use_cache'] = False
```

Comment:
The user could be developing from a specific commit of the SCM (which would get cached), especially if they're developing a patch for someone else's code.

Reply:
Originally I thought this was applying to the source cache. But looking through this (including `Installer._install_task`) I see this is referring to the binary cache. So that makes sense: if a user is developing a package and using a fixed commit that exists in the binary cache, this is needed to avoid pulling that and bypassing the install